presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java

```diff
@@ -313,22 +321,390 @@ public void testHardCpuLimit()
+        assertEquals(q2.getState(), RUNNING);
+        assertEquals(q3.getState(), RUNNING);
+    }
+
+    @Test(timeOut = 10_000)
+    public void testMemoryUsageUpdateAtQueryCompletion()
+    {
+        RootInternalResourceGroup root = new RootInternalResourceGroup("root", (group, export) -> {}, directExecutor());
+        InternalResourceGroup child = root.getOrCreateSubGroup("child");
+
+        // Set the common configurations for the two groups
+        Stream.of(root, child).forEach(group -> {
+            group.setHardConcurrencyLimit(100);
+            group.setMaxQueuedQueries(100);
+            group.setSoftMemoryLimit(new DataSize(3, BYTE));
+        });
+
+        MockManagedQueryExecution q1 = new MockManagedQueryExecutionBuilder().build();
+        child.run(q1);
+        assertEquals(q1.getState(), RUNNING);
+        q1.setMemoryUsage(new DataSize(4, BYTE));
+
+        Stream.of(root, child).forEach(group -> assertWithinMemoryLimit(group, 0));
+        root.updateGroupsAndProcessQueuedQueries();
+        Stream.of(root, child).forEach(group -> assertExceedsMemoryLimit(group, 4));
+
+        // Query completion should reduce the cached memory usage to 0B.
+        q1.complete();
+        Stream.of(root, child).forEach(group -> assertWithinMemoryLimit(group, 0));
+
+        // q2 starts running since usage is within the limit.
+        MockManagedQueryExecution q2 = new MockManagedQueryExecutionBuilder().build();
+        child.run(q2);
+        assertEquals(q2.getState(), RUNNING);
+    }
+
+    /**
+     * A test for correct CPU usage update aggregation and propagation in non-leaf nodes. It uses in a multi
+     * level resource group tree, with non-leaf resource groups having more than one child.
+     */
+    @Test(timeOut = 10_000)
+    public void testRecursiveCpuUsageUpdate()
+    {
+        RootInternalResourceGroup root = new RootInternalResourceGroup("root", (group, export) -> {}, directExecutor());
+        InternalResourceGroup rootChild1 = root.getOrCreateSubGroup("rootChild1");
+        InternalResourceGroup rootChild2 = root.getOrCreateSubGroup("rootChild2");
+        InternalResourceGroup rootChild1Child1 = rootChild1.getOrCreateSubGroup("rootChild1Child1");
+        InternalResourceGroup rootChild1Child2 = rootChild1.getOrCreateSubGroup("rootChild1Child2");
+
+        // Set the same values in all the groups for some configurations
+        Stream.of(root, rootChild1, rootChild2, rootChild1Child1, rootChild1Child2).forEach(group -> {
+            group.setCpuQuotaGenerationMillisPerSecond((new Duration(1, MILLISECONDS)).toMillis());
+            group.setHardConcurrencyLimit(100);
+            group.setMaxQueuedQueries(100);
+        });
+
+        root.setHardCpuLimit(new Duration(16, MILLISECONDS));
+        rootChild1.setHardCpuLimit(new Duration(6, MILLISECONDS));
+        // Setting a higher limit for leaf nodes to make sure they are always in the limit
```

Comment:
Minor nit: I would reorder these so that the leafs are the last lines in this block. This is because I think of full-line comments as usually applying to a block of code.

Reply:
the 3 nodes below the comment are leaves indeed. I added an extra space to separate them from the intermediate ones. I'm not sure if you meant something else.